src/LSST/simulator/SimCore.cpp

```diff
@@ -0,0 +1,165 @@
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <http://www.lsstcorp.org/LegalNotices/>.
+ */
+
+// Class header
+#include "simulator/SimCore.h"
+
+// System headers
+
+// Project headers
+#include "util/Log.h"
+
+using namespace std;
+
+namespace LSST {
+namespace m2cellcpp {
+namespace simulator {
+
+SimCore::SimCore() {
+    _outputPort = control::OutputPortBits::Ptr(new control::OutputPortBits());
+    _inputPort = control::InputPortBits::Ptr(new control::InputPortBits());
+
+    vector<int> motorBreakerInBits = {
+            control::InputPortBits::J1_W9_1_MTR_PWR_BRKR_OK,
+            control::InputPortBits::J1_W9_2_MTR_PWR_BRKR_OK,
+            control::InputPortBits::J1_W9_3_MTR_PWR_BRKR_OK,
+            control::InputPortBits::J2_W10_1_MTR_PWR_BRKR_OK,
+            control::InputPortBits::J2_W10_2_MTR_PWR_BRKR_OK,
+            control::InputPortBits::J2_W10_3_MTR_PWR_BRKR_OK,
+            control::InputPortBits::J3_W11_1_MTR_PWR_BRKR_OK,
+            control::InputPortBits::J3_W11_2_MTR_PWR_BRKR_OK,
+            control::InputPortBits::J3_W11_3_MTR_PWR_BRKR_OK};
+
+    _motorSub = SimPowerSubsystem::Ptr(new SimPowerSubsystem(
+            control::PowerSubsystemConfig::MOTOR,
+            _outputPort, control::OutputPortBits::ILC_MOTOR_POWER_ON,
+            control::OutputPortBits::RESET_MOTOR_BREAKERS,
+            _inputPort, motorBreakerInBits));
+
+    vector<int> commBreakerInBits = {
+            control::InputPortBits::J1_W12_1_COMM_PWR_BRKR_OK,
+            control::InputPortBits::J1_W12_2_COMM_PWR_BRKR_OK,
+            control::InputPortBits::J2_W13_1_COMM_PWR_BRKR_OK,
+            control::InputPortBits::J2_W13_2_COMM_PWR_BRKR_OK,
+            control::InputPortBits::J3_W14_1_COMM_PWR_BRKR_OK,
+            control::InputPortBits::J3_W14_2_COMM_PWR_BRKR_OK};
+
+    _commSub = SimPowerSubsystem::Ptr(new SimPowerSubsystem(
+            control::PowerSubsystemConfig::COMM,
+            _outputPort, control::OutputPortBits::ILC_COMM_POWER_ON,
+            control::OutputPortBits::RESET_COMM_BREAKERS,
+            _inputPort, commBreakerInBits));
+
+    _newOutput = *_outputPort;
+}
+
+
+void SimCore::_simRun() {
+    control::OutputPortBits prevOutput = *_outputPort;
+    while(_simLoop) {
+        util::CLOCK::time_point timestamp = util::CLOCK::now();
+        // Read in new outputPorts set by system.
+        {
+            lock_guard<mutex> lg(_mtx);
+            *_outputPort = _newOutput;
+        }
```

Comment:
Do you need to have a specific content (e.g. `{`+ `}`) here? If so, please help to explain the reason in comment.

Reply:
The brackets are there to limit the scope of the lock_guard to the minimum of what is required. The mutex `_mtx` is locked when the lock_guard `lg` is created. `_mtx` is released at the end of the code block when `lg` falls out of scope. This is a standard thing to do in c++, and is kind of the poster child of RAII [https://en.cppreference.com/w/cpp/language/raii](https://en.cppreference.com/w/cpp/language/raii)